kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -190,6 +193,22 @@ public Transaction build(Engine engine) {
       TableFeatures.validateKernelCanWriteToTable(protocol, metadata, table.getPath(engine));
     }
 
+    Optional<Metadata> newMetadata =
```

Comment:
I wonder if this is the right place where we want to inject this stuff? I think I'm still a bit unclear where we want to do various types of validation + updating protocol/metadatas across TransactionBuilder vs Transaction (also possibly different for various table features?...)

Just curious on your opinion

Reply:
I think the validations should be outside before the txn is created. The only time when the metadata should be updated is for ICT (incase of txn retries due to conflicts). Otherwise everything can be validated outside of txn and avoids passing the context to txn.

Reply:
makes sense to me!